### PR TITLE
Add xcm-simulator

### PR DESCRIPTION
**Pull request by evilrobot-01**
Adds a `xcm-simulator`, based on the [example](https://github.com/paritytech/polkadot/tree/master/xcm/xcm-simulator/example) within the Polkadot repository, with Trappist-specific use cases as tests. Each closure scope in each test can be thought of as a block on the corresponding chain, with messages being dispatched to target chains via the mock message queue as it goes out of scope.

It effectively translates the integration tests in https://github.com/paritytech/trappist/pull/76 into Rust with the following main benefits:
- enables interactive debugging with breakpoints, allowing stepping through code to see exactly what is going on as messages are processed across chains
- uses existing runtime config as much as possible, simulating actual runtime interactions as best as possible
- shows xcm tracing output in test console for further additional insight

The tests can simply be run from the command line as follows, assuming one is in the root folder of the project, but the real value is running the tests with a debugger and breakpoints to explore how the instructions are handled by the `xcm-executor`. Note also that the crate is excluded from the workspace, effectively making it 'opt-in'.

```
cd xcm-simulator; cargo test; cd ..
```

Potential next steps:
- implement a helper function for determining fees, better demonstrating the relationships between xcm weights/fees
- add GH action to run when creating PRs, ensuring no regressions in cross-chain functionality

Any other suggestions for improvement welcome!

### CODE REVIEW

**Review by stiiifff — COMMENTED** [review]
Great work @evilrobot-01, this is a super valuable addition to the repo.
Can you add instructions in the README on how to run the tests ?
Also, as you mentioned it, it would be nice to run these as part of a CI task.

**Comment by evilrobot-01** [comment]
Updated the PR description with a crude way to run the tests. Will update the readme and add a separate GH action for now, which can always be incorporated into the main CI flow as/when required.

File: xcm-simulator/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+			});
+			ext
+		},
+	}
+}
+
+const A_PARA_ID: u32 = 1;
+decl_test_parachain! {
+	// Some generic parachain
+	pub struct ParaA {
```
Comment (bernardoaraujor) [inline xcm-simulator/src/lib.rs:139]: what's the role of `ParaA`?

**Comment by stiiifff** [comment]
Fixes #103

File: README.md [hunk 2]
```diff
@@ -83,6 +83,17 @@ Then, start the **Trappist** playground with:
...
 
+### XCM Simulator
+The [XCM simulator](./xcm-simulator) can be used to further explore XCM message execution across the various runtimes used by Trappist. 
+Each Trappist use case is written as a test, allowing interactive debugging/exploration of message flows and instruction execution.
+Each `execute_with` closure scope within a test can be considered as a block on the corresponding chain, with messages being dispatched to the destination chains via a mock message queue as the closure goes out of scope.
+All XCM-specific traces from the interactions are also collected in a single place for easier inspection. 
+
+You can run all tests, including the tracing output, with:
+```
+cd xcm-simulator && cargo nextest run tests:: --no-capture --release
```
Comment (bernardoaraujor) [inline README.md:94]: maybe add some contextual note about `cargo-nextest`?

or just instruct audience to use vanilla `cargo test`

File: xcm-simulator/src/lib.rs (hunk 1 above)
Comment (bernardoaraujor) [reply xcm-simulator/src/lib.rs:139]: I'd give the variable a more insightful name, and also a more detailed description in the comments.

File: README.md (hunk 2 above)
Comment (evilrobot-01) [reply README.md:94]: Sorry, that was a copy-paste error and shouldn't have been pushed yet. Think it got caught up in the CI push I did to test that on GitHub. Will get it corrected to standard `cargo test`.

Thanks!

File: README.md [hunk 3]
```diff
@@ -83,6 +83,17 @@ Then, start the **Trappist** playground with:
 ./zombienet-linux -p native spawn xcm-playground.toml
 ```
 
+### XCM Simulator
```
Comment (bernardoaraujor) [inline README.md:86]:
```suggestion
### Integration Tests
[parachains-integration-tests](https://github.com/paritytech/parachains-integration-tests) is a tool meant for XCM message execution in a locally spawned network. Tests are written as YAML files and converted into [Mocha](https://mochajs.org/) tests with [Chai](https://www.chaijs.com/) assertions.

The [`integration-tests`] directory has tests on Trappist use cases and instructions on how to run them.

### XCM Simulator
```

File: README.md [hunk 4]
```diff
@@ -83,6 +83,17 @@ Then, start the **Trappist** playground with:
...
+The [XCM simulator](./xcm-simulator) can be used to further explore XCM message execution across the various runtimes used by Trappist. 
+Each Trappist use case is written as a test, allowing interactive debugging/exploration of message flows and instruction execution.
```
Comment (bernardoaraujor) [inline README.md:88]:
```suggestion
Each Trappist use case is written as a Rust unit test, allowing interactive debugging/exploration of message flows and instruction execution.
```

File: xcm-simulator/src/lib.rs (hunk 1 above)
Comment (evilrobot-01) [reply xcm-simulator/src/lib.rs]: Renamed to `Template` to better describe intention, along with improved comments

**Comment by evilrobot-01** [comment]
Updated the readme, thanks to additional information from @bernardoaraujor 🙌, and then also added a separate CI workflow which outputs the test results as an [artifact](https://github.com/paritytech/trappist/pull/102/checks?check_run_id=10560568569). I had hoped to also include the `xcm-executor` tracing output as further demonstration of what is available from the simulations, but `nextest` doesnt currently support this for passing tests so I have opened a feature request issue there.

PS - `nextest` used in CI as it allows converting results into junit format for nicer reporting using [test-reporter](https://github.com/dorny/test-reporter). Happy to hear of any better alternatives, haven't explored the GH test reporting landscape for a few months.

**Comment by stiiifff** [comment]
@evilrobot-01 Super cool, I love the test report. Should we add a badge to the README with the latest "XCM tests" status ?

**Comment by evilrobot-01** [comment]
Sure, any preference on location?

**Review by bernardoaraujor — APPROVED** [review]

**Pull request merged**